core/src/main/java/com/rudderstack/android/sdk/core/RudderDeviceModeTransformationManager.java

```diff
@@ -60,39 +62,53 @@ public void run() {
                                 }
-                                String requestJson = createDeviceTransformPayload(messageIds, messages);
+                                ArrayList<String> transformationEnabledDestinationIds = getTransformationEnabledDestinationIds(messages);
+                                String requestJson = createDeviceTransformPayload(messageIds, messages, transformationEnabledDestinationIds);
                                 RudderLogger.logDebug(String.format(Locale.US, "DeviceModeTransformationManager: TransformationProcessor: Payload: %s", requestJson));
                                 RudderLogger.logInfo(String.format(Locale.US, "DeviceModeTransformationManager: TransformationProcessor: EventCount: %d", messageIds.size()));
-                                Result result = rudderNetworkManager.sendNetworkRequest(requestJson, addEndPoint(dataResidencyManager.getDataPlaneUrl(), TRANSFORMATION_ENDPOINT), RequestMethod.POST, true, true);
+                                Result result = rudderNetworkManager.sendNetworkRequest(requestJson, addEndPoint(dataResidencyManager.getDataPlaneUrl(), TRANSFORMATION_ENDPOINT), RequestMethod.POST, false, true);
                                 if (result.status == NetworkResponses.WRITE_KEY_ERROR) {
                                     RudderLogger.logDebug("DeviceModeTransformationManager: TransformationProcessor: Wrong WriteKey. Aborting");
                                     break;
+                                } else if (result.status == NetworkResponses.NETWORK_UNAVAILABLE) {
+                                    RudderLogger.logDebug("DeviceModeTransformationManager: TransformationProcessor: Network unavailable. Aborting");
+                                    break;
                                 } else if (result.status == NetworkResponses.ERROR) {
-                                    RudderLogger.logDebug("DeviceModeTransformationManager: TransformationProcessor: Retrying in " + Math.abs(deviceModeSleepCount - config.getSleepTimeOut()) + "s");
-                                    try {
-                                        Thread.sleep(Math.abs(deviceModeSleepCount - config.getSleepTimeOut()) * 1000L);
-                                    } catch (Exception e) {
-                                        RudderLogger.logError(e);
-                                        Thread.currentThread().interrupt();
+                                    int delay = Math.min((1 << retryCount) * 500, MAX_DELAY); // Exponential backoff
+                                    if (retryCount++ == MAX_RETRIES) {
+                                        retryCount = 0;
+                                        deviceModeSleepCount = 0;
+                                        rudderDeviceModeManager.dumpOriginalEvents(gson.fromJson(requestJson, TransformationRequest.class), true);
```

Comment:
We are deserializing same events in line 72 and here. Why can't we deserialize once and dump those?

Reply:
In line 72, we are creating the list of `destinationIds`, right? refer [here](https://github.com/rudderlabs/rudder-sdk-android/blob/f4dbcfa3cc626db146ae39fe14b7337448e4ab67/core/src/main/java/com/rudderstack/android/sdk/core/RudderDeviceModeTransformationManager.java#L72C33-L72C138):
`ArrayList<String> transformationEnabledDestinationIds = getTransformationEnabledDestinationIds(messages);`

And in line [88](https://github.com/rudderlabs/rudder-sdk-android/blob/f4dbcfa3cc626db146ae39fe14b7337448e4ab67/core/src/main/java/com/rudderstack/android/sdk/core/RudderDeviceModeTransformationManager.java#L88), we are deserialising the message.

Both are different activities, right?

Please LMK if you mean something else.